design-proposals/vm-memory-snapshot.md

```diff
@@ -0,0 +1,151 @@
+        devices:
+          disks:
+          - disk:
+              bus: virtio
+            name: containerdisk
+          - disk:
+              bus: virtio
+            name: cloudinitdisk
+        resources:
+          requests:
+            memory: 512Mi
+      terminationGracePeriodSeconds: 0
+      volumes:
+      - containerDisk:
+          image: registry:5000/kubevirt/fedora-with-test-tooling-container-disk:devel
+        name: containerdisk
+      - memoryDump:
+          persistentVolumeClaim:
+              claimName: memory-dump
+        name: disk1
+      - cloudInitNoCloud:
+          userData: |-
+            #cloud-config
+            password: fedora
+            chpasswd: { expire: False }
+        name: cloudinitdisk
+```
+
+This can be added before creating the VM, or if the VM already exists need to add and restart the VM to make the pvc mounted.
+
+#### Add memory dump volume as a hotplug
+
+Use the hot-plug API to add a memory dump volume by adding a flag to the `virtctl addVolume` command:
+
+`$ virtctl addvolume my-vm --volume-name=memory-dump --memoryDumpVolume`
+
+This will hotplug the pvc `memory-hotplug` as a memoryDumpVolumeSource to the virt launcher pod, without mounting it to the VMI.
+De-facto the command will call the addvolume subresource as done in hot-plug, will add the volume to the VM volume spec and to the VMI volumes Spec which should add the volume to the VMI volume status as a memoryDump volume, which should trigger the mount to the virt-launcher pod.
+
+### Get memory dump
+#### trigger memory dump
+After adding the pvc to which the memory dump should be saved to, the user can trigger the memory dump in two ways:
+
+1. Run a virtctl command to get a memory dump. This new API will look as follows:
+
+`$ virtctl memoryDump my-vm`
+
+2. Get a memory dump during Online VMSnapshot. This will be done by adding to the VMSnapshot yaml a parameter to indicate that we also want a memory dump. For example:
+
+```yaml
+apiVersion: snapshot.kubevirt.io/v1alpha1
+kind: VirtualMachineSnapshot
+metadata:
+  name: vm-snap
+spec:
+  source:
+    apiGroup: kubevirt.io
+    kind: VirtualMachine
+    name: my-vm
+  includeMemory: true
+```
+
+#### How to handle it
+The trigger of memory dump will call a vmi subresource that will trigger `virDomainCoreDump` command with `VIR_DUMP_MEMORY_ONLY` flag to be executed by virt-launcher. Since this can take some time it will be an asynchronous operation. The memoryDump volume in the VMI-volume-status will change to `memoryDumpInProgress` and when completed it will change to `memoryReady` to indicate that the memory dump is ready to use.
+
+If the memory dump was triggered by memoryDump command then the user will be able to watch the progress by looking at the VMI memorydump volume status.
+
+In the Online VMSnapshot the memoryDump will be triggered after the vmfreeze. Taking into account that for a big memory dump it can take more time for the snapshot to be completed.
```

Comment:
What is vmfreeze?  Is the VM paused? Or is it an fsfreeze?

Reply:
is it fsfreeze. for the dump command I understand that the vm will be suspended and after unsuspended hence it will take care for the consistency of the dump and the wrapping fsfreeze will take care for the consistency of the dump with the disks.

Reply:
I think we're are aligned, but can you add the expectation that the VM is suspended/paused to the doc?